ibis/bigquery/client.py

```diff
@@ -0,0 +1,207 @@
+import re
+
+import ibis
+import ibis.expr.types as ir
+import ibis.expr.datatypes as dt
+from ibis.bigquery import compiler as comp
+import toolz
+import google.datalab as gd
+import google.datalab.bigquery as bq
+from ibis.client import (Database, SQLClient)
+
+
+_fully_qualified_re = re.compile('(.*)\.(.*)|(.*)')
+
+
+# pip install git+https://github.com/googledatalab/pydatalab
+def _parse_bq_maybe_qualified_re(name):
+    (dataset, table0, table1) = _fully_qualified_re.match(name).groups()
+    table = table0 or table1
+    return (dataset, table)
+
+
+def _ensure_split(table_id, dataset_id):
+    (_dataset_id, table_id) = _parse_bq_maybe_qualified_re(table_id)
+    assert (dataset_id == _dataset_id) or not (dataset_id and _dataset_id)
+    dataset_id = dataset_id or _dataset_id
+    return (table_id, dataset_id)
+
+
+def _bq_table_obj_to_string(table):
+    (_, dataset_id, table_id, _) = table.name
+    return dataset_id + '.' + table_id
+
+
+def _bq_make_context(project_id=None):
+    context = gd.Context.default()
+    if context.project_id != project_id:
+        context.set_project_id(project_id)
+    return context
+
+
+@toolz.memoize
+def _bq_get_context(project_id):
+    return _bq_make_context(project_id)
+
+
+def _bq_get_datasets(project_id):
+    return bq.Datasets(_bq_get_context(project_id))
+
+
+def _bq_get_dataset(dataset_id, project_id):
+    return bq.Dataset(dataset_id, _bq_get_context(project_id))
+
+
+def _bq_get_table(table_id, dataset_id, project_id):
+    (table_id, dataset_id) = _ensure_split(table_id, dataset_id)
+    return bq.Table(dataset_id + '.' + table_id, _bq_get_context(project_id))
+
+
+def _bq_get_schema(table_id, dataset_id, project_id):
```

Comment:
Why not keep a `_context` on the client (I see that you have a property based on the `project_id`) and use the `bq.Table`/`bq.Dataset(s)` objects directly instead of having these extra 1-2 line functions?